PCGen-base/code/src/java/pcgen/base/format/ArrayFormatManager.java

```diff
@@ -58,105 +72,123 @@
 	/**
 	 * Converts the instructions into an array of objects. The objects referred
 	 * to in the instructions should be separated by the separator provided at
 	 * construction of this ArrayFormatManager.
 	 */
+	@SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
 	@Override
 	public T[] convert(String instructions)
 	{
-		Class<T> componentClass = componentManager.getManagedClass();
 		if ((instructions == null) || instructions.isEmpty())
 		{
-			@SuppressWarnings("unchecked")
-			T[] toSet = (T[]) Array.newInstance(componentClass, 0);
-			return toSet;
+			return ArrayUtilities.buildEmpty(componentManager.getManagedClass());
 		}
-		if (!hasValidSeparators(instructions))
+		Class<T> managedClass = componentManager.getManagedClass();
+		if (componentManager instanceof DispatchingFormatManager)
 		{
-			throw new IllegalArgumentException(
-				"Poorly formatted instructions (bad separator location): "
-					+ instructions);
-		}
-		String[] items = instructions.split(Character.toString(separator));
-		@SuppressWarnings("unchecked")
-		T[] toSet = (T[]) Array.newInstance(componentClass, items.length);
-		for (int i = 0; i < items.length; i++)
-		{
-			T obj = componentManager.convert(items[i]);
-			toSet[i] = obj;
+			@SuppressWarnings("unchecked")
+			DispatchingFormatManager<T> dfm =
+					(DispatchingFormatManager<T>) componentManager;
+			String[] groups = splitInstructions(instructions, groupSep);
+			T[] returnValue = null;
+			for (String group : groups)
+			{
+				T[] converted = dfm.convertViaDispatch(
+					fm -> new ArrayFormatManager<>(fm, groupSep, listSep), group);
+				returnValue =
+						ArrayUtilities.mergeArray(managedClass, returnValue, converted);
+			}
+			return returnValue;
 		}
-		return toSet;
-	}
-
-	private boolean hasValidSeparators(String value)
-	{
-		//assume not empty due to checks on instructions
-		return (value.charAt(0) != separator)
-			&& (value.charAt(value.length() - 1) != separator)
-			&& (!value.contains(String.valueOf(new char[]{separator, separator})));
+		return convertInternal(componentManager::convert, instructions, managedClass);
 	}
 
 	/**
 	 * Converts the instructions into an Indirect array of objects. The objects
 	 * referred to in the instructions should be separated by the separator
 	 * provided at construction of this ArrayFormatManager.
 	 */
+	@SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
 	@Override
 	public Indirect<T[]> convertIndirect(String instructions)
-	{
-		/*
-		 * Common code with convertObjectContainer may be a surprise, but
-		 * consider that the instructions may not just be a,b,c. They could be
-		 * things like: a,GROUP=b. The GROUP= will return more than one object,
-		 * so on everything here, we want to check ObjectContainer to get the
-		 * full set of items.
-		 */
-		return convertInstructions(instructions);
-	}
-
-	private ArrayIndirect convertInstructions(String instructions)
 	{
 		if ((instructions == null) || instructions.isEmpty())
+		{
```

Comment:
Again not sure, but I thing Arrays and Collections would help here

Reply:
Same as above